src/transformStateWithClones.js

```diff
@@ -7,7 +7,45 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = addProperties(currentState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        currentState = removeProperties(currentState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        currentState = clearProperties();
```

Comment:
[CODE STYLE] - don't mutate object or arrays - it will cause unexpected results later on. While the code correctly makes copies of the state, the naming of the copy should be improved to better reflect its purpose. Instead of `currentState`, consider using `stateCopy` or another descriptive name that indicates it's a copy of the state.